internal/app/cache/cache_test.go

```diff
@@ -0,0 +1,136 @@
+		Identifier: "identifier_A",
+	},
+}
+
+func TestExists_EmptyCache(t *testing.T) {
+	cache, err := NewCache(10, 1048576, time.Second*60, onEvict)
+	assert.NoError(t, err)
+
+	assert.False(t, cache.Exists(testKeyA))
+}
+
+func TestAddRequestAndFetch(t *testing.T) {
+	cache, err := NewCache(10, 1048576, time.Second*60, onEvict)
+	assert.NoError(t, err)
+
+	// Simulate cache miss and setting of new request.
+	response, err := cache.Fetch(testKeyA)
+	assert.EqualError(t, err, "No value found for key: key_A")
+	assert.Nil(t, response)
+	isStreamOpen, err := cache.AddRequest(testKeyA, testRequest)
+	assert.NoError(t, err)
+	assert.True(t, isStreamOpen)
+	time.Sleep(1 * time.Millisecond)
+	assert.True(t, cache.Exists(testKeyA))
+	response, err = cache.Fetch(testKeyA)
+	assert.NoError(t, err)
+	assert.Nil(t, response)
+}
+
+func TestSetResponseAndFetch(t *testing.T) {
+	cache, err := NewCache(10, 1048576, time.Second*60, onEvict)
+	assert.NoError(t, err)
+
+	// Simulate cache miss and setting of new response.
+	response, err := cache.Fetch(testKeyA)
+	assert.EqualError(t, err, "No value found for key: key_A")
+	assert.Nil(t, response)
+	requests, err := cache.SetResponse(testKeyA, testResponse)
+	assert.NoError(t, err)
+	assert.Nil(t, requests)
+	time.Sleep(1 * time.Millisecond)
+	response, err = cache.Fetch(testKeyA)
+	assert.NoError(t, err)
+	assert.Equal(t, testResponse, *response)
+}
+
+// This test demonstrates behavior unique to ristretto caching, i.e. if Set is applied on a new key, it may take
+// a few milliseconds after the call returns, but if the key already exists in the cache, the update is done instantly.
+func TestAddRequestAndSetResponse(t *testing.T) {
+	cache, err := NewCache(10, 1048576, time.Second*60, onEvict)
+	assert.NoError(t, err)
+
+	isStreamOpen, err := cache.AddRequest(testKeyA, testRequest)
+	assert.NoError(t, err)
+	assert.True(t, isStreamOpen)
+	time.Sleep(1 * time.Millisecond)
+	isStreamOpen, err = cache.AddRequest(testKeyA, testRequest)
+	assert.NoError(t, err)
+	assert.True(t, isStreamOpen)
+
+	requests, err := cache.SetResponse(testKeyA, testResponse)
+	assert.NoError(t, err)
+	assert.Equal(t, 2, len(requests))
+	assert.Equal(t, testRequest, *requests[0])
+	assert.Equal(t, testRequest, *requests[1])
+
+	response, err := cache.Fetch(testKeyA)
+	assert.NoError(t, err)
+	assert.Equal(t, testResponse, *response)
+}
+
+func TestTTL(t *testing.T) {
+	cache, err := NewCache(10, 1048576, time.Second*1, onEvict)
+	assert.NoError(t, err)
+	_, err = cache.AddRequest(testKeyA, testRequest)
+	assert.NoError(t, err)
+	time.Sleep(1 * time.Millisecond)
+	assert.True(t, cache.Exists(testKeyA))
+	time.Sleep(1 * time.Second)
+	assert.False(t, cache.Exists(testKeyA))
+}
+
+func TestMemoryOverflow(t *testing.T) {
+	cache, err := NewCache(10, 40, time.Second*60, onEvict)
+	assert.NoError(t, err)
+	_, err = cache.AddRequest(testKeyA, testRequest)
+	assert.NoError(t, err)
+	time.Sleep(1 * time.Millisecond)
```

Comment:
Echoing what @eapolinario said, I would avoid putting these sleeps in the tests. Do we need them right now?

Reply:
https://github.com/envoyproxy/xds-relay/blob/e385cbec086fdc08ac8c46bfcefcc4ff9fb993a8/internal/app/cache/cache_test.go#L84-L107 demonstrates the ristretto cache behavior, in which without the sleep between AddRequest calls (https://github.com/envoyproxy/xds-relay/blob/e385cbec086fdc08ac8c46bfcefcc4ff9fb993a8/internal/app/cache/cache_test.go#L93), only 1 request will be associated with that key (https://github.com/envoyproxy/xds-relay/blob/e385cbec086fdc08ac8c46bfcefcc4ff9fb993a8/internal/app/cache/cache_test.go#L100 would fail). I'm not sure why the cache drops the request, even with the lock.